harvester/records.py

```diff
@@ -0,0 +1,225 @@
+"""harvester.parse"""
+
+import base64
+import io
+import logging
+import time
+import xml.etree.ElementTree as ET
+
+import jsonlines  # type: ignore[import]
+import pandas as pd
+import smart_open  # type: ignore[import]
+
+from harvester.wacz import WACZClient
+
+logger = logging.getLogger(__name__)
+
+
+class CrawlRecordsParser:
+    """Class to facilitate parsing records from completed crawls.
+
+    This parser is designed to parse a WACZ file, a compressed archive of a web crawl, and
+    generate records for each website crawled.
+    """
+
+    def __init__(self, wacz_filepath: str):
+        self.wacz_filepath = wacz_filepath
+        self._websites_records: CrawlRecords | None = None
+
+    def generate_records(
+        self,
+        *,
+        urls_file: str | None = None,
+        previous_sitemap_urls_file: str | None = None,
+    ) -> "CrawlRecords":
+        """Generate dataframe of records from websites crawled in WACZ file.
+
+        This combines data from the crawl (e.g. URL, WARC file), and the full,
+        rendered HTML from the website itself.
+
+        Args:
+            urls_file: Text file with URLs found in this crawl
+                - used for generating action="delete" records
+            previous_sitemap_urls_file: Text file with URLs found in last crawl
+                - used for generating action="delete" records
+        """
+        logger.info("Generating records from crawl")
+
+        # if records already generated, re-use
+        if self._websites_records:
+            return self._websites_records
+
+        all_records = []
+
+        # add records from the crawl (new or modified sites)
+        with WACZClient(self.wacz_filepath) as wacz_client:
+            t0 = time.time()
+            row_count = len(wacz_client.html_websites_df)
+
+            # loop through HTML pages captured in crawl
+            for i, row in enumerate(wacz_client.html_websites_df.itertuples(), 1):
+                if i % 100 == 0:  # pragma: no cover
+                    logger.debug(
+                        "%d/%d records parsed, %.2fs", *(i, row_count, time.time() - t0)
+                    )
+                    t0 = time.time()
+
+                # init record dictionary with data already known from WACZ files
+                record = {
+                    "url": row.url,
+                    "status": "active",
+                    "cdx_warc_filename": row.filename,
+                    "cdx_title": row.title,
+                    "cdx_offset": row.offset,
+                    "cdx_length": row.length,
+                }
+
+                # add base64 encoded full HTML
+                html_content = wacz_client.get_website_content(
+                    str(row.filename),
+                    str(row.offset),
+                    decode=False,
+                )
+                record["html_base64"] = base64.b64encode(html_content).decode()  # type: ignore[arg-type]
```

Comment:
This is where we base64 encode and store the page HTML in the output record.  This replaces a considerable amount of logic and complexity where we parsed metadata tags from the HTML.